### ISSUE TITLE
Unable to build godot_remote module for macOS

### ISSUE THREAD

**Issue by atonamy** [DmitriySalnikov/GodotRemote#2]
I'm experience compilation errors when I trying to build this module for macOS in Godot3.3 source:
`scons platform=osx arch=x86_64 tools=yes module_godot_remote_enabled=yes --jobs=$(sysctl -n hw.logicalcpu)`
> scons: Reading SConscript files ...
> Building for macOS 10.9+, platform x86-64.
> IsADirectoryError: [Errno 21] Is a directory: '/Users/archie/Projects/GameDev/godot/modules/godot_remote/openh264/':
>   File "/Users/archie/Projects/GameDev/godot/SConstruct", line 637:
>     SConscript("modules/SCsub")
>   File "/usr/local/Cellar/scons/4.1.0.post1/libexec/lib/python3.9/site-packages/SCons/Script/SConscript.py", line 654:
>     return method(*args, **kw)
>   File "/usr/local/Cellar/scons/4.1.0.post1/libexec/lib/python3.9/site-packages/SCons/Script/SConscript.py", line 591:
>     return _SConscript(self.fs, *files, **subst_kw)
>   File "/usr/local/Cellar/scons/4.1.0.post1/libexec/lib/python3.9/site-packages/SCons/Script/SConscript.py", line 280:
>     exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
>   File "/Users/archie/Projects/GameDev/godot/modules/SCsub", line 21:
>     SConscript(name + "/SCsub")  # Built-in.
>   File "/usr/local/Cellar/scons/4.1.0.post1/libexec/lib/python3.9/site-packages/SCons/Script/SConscript.py", line 654:
>     return method(*args, **kw)
>   File "/usr/local/Cellar/scons/4.1.0.post1/libexec/lib/python3.9/site-packages/SCons/Script/SConscript.py", line 591:
>     return _SConscript(self.fs, *files, **subst_kw)
>   File "/usr/local/Cellar/scons/4.1.0.post1/libexec/lib/python3.9/site-packages/SCons/Script/SConscript.py", line 280:
>     exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
>   File "/Users/archie/Projects/GameDev/godot/modules/godot_remote/SCsub", line 41:
>     lib_utils.prepare_h264(env)
>   File "/Users/archie/Projects/GameDev/godot/modules/godot_remote/lib_utils.py", line 194:
>     copyfile(cwd + old_lib, new_dir + lib_openh264)
>   File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/shutil.py", line 264:
>     with open(src, 'rb') as fsrc, open(dst, 'wb') as fdst:

If I try to use` scons platform=osx arch=x86_64 tools=yes module_godot_remote_enabled=yes godot_remote_h264_enabled=no --jobs=$(sysctl -n hw.logicalcpu) `

I getting another error

> scons: Reading SConscript files ...
> Building for macOS 10.9+, platform x86-64.
> Checking for C header file mntent.h... (cached) no
> scons: done reading SConscript files.
> scons: Building targets ...
> ...
>  [ 27%] Compiling ==> thirdparty/opus/silk/float/autocorrelation_FLP.c
> modules/godot_remote/GRServer.cpp:754:67: error: conversion from 'std::__1::vector<std::__1::shared_ptr<UdpBroadcast>>::size_type' (aka 'unsigned long') to 'const Variant' is ambiguous
>         _log("UDP Broadcast thread inited with " + str(available_sockets.size()) + " available addresses", LogLevel::LL_DEBUG);

> [ 27%] modules/godot_remote/GRServer.cpp:1503:21: error: conversion from 'std::__1::vector<String>::size_type' (aka 'unsigned long') to 'const Variant' is ambiguous
>                                         _log(str(files.size()) + " files added to custom input PCK", LogLevel::LL_NORMAL);
> [ 28%] Compiling ==> thirdparty/embree/common/sys/filename.cpp
> modules/godot_remote/register_types.cpp:64:26: error: unknown type name 'GRStreamEncoderH264'; did you mean 'GRStreamEncoder'?
>         ClassDB::register_class<GRStreamEncoderH264>();
> 23 warnings and 2 errors generated.
> scons: *** [modules/godot_remote/GRServer.osx.tools.x86_64.o] Error 1
> 45 warnings and 1 error generated.
> scons: *** [modules/godot_remote/register_types.osx.tools.x86_64.o] Error 1
> scons: building terminated because of errors.

Is anyway to make it compilable for OSX?

**Comment by DmitriySalnikov** [comment]
> I'm experience compilation errors when I trying to build this module for macOS in Godot3.3 source:
> `scons platform=osx arch=x86_64 tools=yes module_godot_remote_enabled=yes --jobs=$(sysctl -n hw.logicalcpu)`

Yeap, there are no pre-compiled libraries for h264 for osx. I can add them from [here](https://github.com/cisco/openh264/releases/tag/v2.1.1), but I'm not sure it will help because I don't have a macOS device

> modules/godot_remote/GRServer.cpp:754:67: error: conversion from 'std::__1::vector<std::__1::shared_ptr>::size_type' (aka 'unsigned long') to 'const Variant' is ambiguous
> _log("UDP Broadcast thread inited with " + str(available_sockets.size()) + " available addresses", LogLevel::LL_DEBUG);

hmm, that's strange.

> modules/godot_remote/register_types.cpp:64:26: error: unknown type name 'GRStreamEncoderH264'; did you mean 'GRStreamEncoder'?

and this is probably a bug across all platforms.

Also after these errors, there should be another problem with linking libjpeg-turbo. I build this library myself, and I need macOS for that.

**Comment by DmitriySalnikov** [comment]
@atonamy, can you check the latest changes? c253a9842c9ce1c2eae5f479da278bf919211df0

**Comment by atonamy** [comment]
> @atonamy, can you check the latest changes? [c253a98](https://github.com/DmitriySalnikov/GodotRemote/commit/c253a9842c9ce1c2eae5f479da278bf919211df0)

Yes it's working now. Even ` scons platform=osx arch=x86_64 tools=yes module_godot_remote_enabled=yes --jobs=$(sysctl -n hw.logicalcpu)` works. Except I got bunch of warnings.

Regarding functionality I only able to connect by WIFI, ADB is not connecting (maybe I miss something), AUTO is also not working. 

You can use OSX VM from Sosumi to run MacOS on Linux host and test all scenarios regarding build and execution correctness  of module. 
https://www.youtube.com/watch?v=QWZ_LjzT39k&t=92s

Highly appreciate if you will add prebuild binaries for MacOS as well on itch.io including binaries for Apple Silicon.

**Comment by DmitriySalnikov** [comment]
> Except I got bunch of warnings.

so it was on linux. Now I've disabled them. 62a1672c2830e08d4588cfb166b3ba76ecb0045c

> Regarding functionality I only able to connect by WIFI, ADB is not connecting (maybe I miss something), AUTO is also not working.

Auto mode is based on UDP broadcasting, but broadcasting is not possible on macOS as I understand it. [source](https://stackoverflow.com/questions/33072128/udp-broadcast-via-sendto-returns-cant-assign-requested-address-on-os-x-but)

> You can use OSX VM from Sosumi to run MacOS on Linux host and test all scenarios regarding build and execution correctness of module.
> https://www.youtube.com/watch?v=QWZ_LjzT39k&t=92s

My primary operating system is Windows and I have already spent several hours setting up VM and testing everything on it.

> Highly appreciate if you will add prebuild binaries for MacOS as well on itch.io including binaries for Apple Silicon.

The build for macos is very slow for me (1-1.5 hours per architecture), while all other platforms are built in 1.5-2 hours. This is an open source project that does not bring any money to me and I do not want to spend time building for a platform that does not work correctly.
Someday, if I rework the build process, I'll probably add macos.
